Docs: New guide for importing large datasets (#18852)
supabase/supabase: apps/docs/pages/guides/database/large-datasets.mdx

import Layout from '~/layouts/DefaultGuideLayout'

export const meta = {
  id: 'database',
  title: 'Importing into Supabase',
  description: 'Importing Large Datasets into Supabase.',
  subtitle: 'Importing Large Datasets into Supabase.',
  sidebar_label: 'FUNDAMENTALS',
}

Importing large datasets requires careful planning and execution to ensure the process is efficient, reliable, and doesn't disrupt your application. In this guide, we cover various best practices and techniques for importing datasets, while also highlighting common pitfalls to avoid.

> ## **You have multiple options for importing your data into Supabase**

Reviewer: For headings, I don't think you'll need `**` - they will be formatted accordingly by default

> ## **You have multiple options for importing your data into Supabase**

Reviewer: ```suggestion
## You have multiple options for importing your data into Supabase
```

### Option 1: CSV Import via Supabase Dashboard

>     Supabase dashboard provides a user-friendly way to import data. However, for very large datasets, this method may not be the most efficient choice, given the size limit is 100MB. It's generally better suited for smaller datasets and quick data imports. Consider using alternative methods like pgloader for large-scale data imports.

Reviewer: There seems to be a bit of unnecessary space at the beginning here

```suggestion
Supabase dashboard provides a user-friendly way to import data. However, for very large datasets, this method may not be the most efficient choice, given the size limit is 100MB. It's generally better suited for smaller datasets and quick data imports. Consider using alternative methods like pgloader for large-scale data imports.
```

    - Navigate to the relevant table in the [Table Editor.](https://supabase.com/dashboard/project/_/editor)
    - Click on “Insert” then choose "Import Data from CSV" and follow the on-screen instructions to upload your CSV file.

### Option 2: Bulk Import using pgloader

>     [pgloader](https://pgloader.io/) is a powerful tool for efficiently importing data into a PostgreSQL database that supports a wide range of source database engines, including MySQL and MS SQL.

Reviewer: ```suggestion
[pgloader](https://pgloader.io/) is a powerful tool for efficiently importing data into a PostgreSQL database that supports a wide range of source database engines, including MySQL and MS SQL.
```

>     You can use it in conjunction with Supabase by following these steps:

Reviewer: There's a bunch of empty spaces at the beginning of these sentences it seems, could you remove them please?

    1. Install pgloader on your local machine or a server. For more info, you can refer to the official pgloader installation page available at: https://pgloader.readthedocs.io/en/latest/install.html

        ```jsx
        $ apt-get install pgloader
        ```

    2. Create a configuration file that specifies the source data and the target Supabase database (e.g., config.load).

        Here's an example configuration file:
        ```sql
        LOAD DATABASE
            FROM sourcedb://USER:PASSWORD@HOST/SOURCE_DB
            INTO postgres://postgres:password@db.xxxx.supabase.co:6543/postgres
        ALTER SCHEMA 'public' OWNER TO 'postgres';

>         set wal_buffers = '64MB', max_wal_senders = 0, statement_timeout = 0, work_mem to '2GB';

Reviewer: I think it's important to clarify these parameters, I mean most parameters like schema, the URLs etc are quite clear - but I wouldn't know why we're setting wal_buffers, max_wal_senders and work_mem

        ```

        Customize the source and Supabase database URL and options to fit your specific use case.

    3. Run pgloader with the configuration file.

        ```jsx
        pgloader config.load
        ```

    For databases using the Postgres engine, we recommend using the [pg_dump](https://www.postgresql.org/docs/current/app-pgdump.html) and [psql](https://www.postgresql.org/docs/current/app-psql.html) command line tools.

### Option 3: Using Postgres Copy command

    Read more about [Bulk data loading.](https://supabase.com/docs/guides/database/tables#bulk-data-loading)

### Option 4: Using the Supabase API

    The Supabase API allows you to programmatically import data into your tables. You can use various client libraries to interact with the API and perform data import operations. This approach is useful when you need to automate data imports, and it gives you fine-grained control over the process. Refer to our [API guide](https://supabase.com/docs/guides/api) for more details.

>     <Admonition type="note">

Reviewer: Could use an empty line between the previous paragraph and <Admonition

    When importing data via the Supabase API, it's advisable to refrain from bulk imports. This helps ensure a smooth data transfer process and prevents any potential disruptions.

    Read more about [Rate Limiting, Resource Allocation, & Abuse Prevention.](https://supabase.com/docs/guides/platform/going-into-prod#rate-limiting-resource-allocation--abuse-prevention)
    </Admonition>

## **Before you dive into the import process, it's essential to plan and prepare**

### 1. Statement Timeouts

By default, Supabase enforces query statement timeouts to ensure fair resource allocation and prevent long-running queries from affecting the overall system. When importing large datasets, you may encounter timeouts. To address this:

> - **Increase the Statement Timeout**: You can adjust the statement timeout for your session or connection to accommodate longer-running queries. Be cautious when doing this, as excessively long queries can negatively impact system performance. You can read more about Statement Timeouts here: https://supabase.com/docs/guides/database/postgres/configuration

Reviewer: ```suggestion
- **Increase the Statement Timeout**: You can adjust the statement timeout for your session or connection to accommodate longer-running queries. Be cautious when doing this, as excessively long queries can negatively impact system performance. Read more about [Statement Timeouts](https://supabase.com/docs/guides/database/postgres/configuration).
```

### 2. Disk Size Considerations

> Large datasets consume disk space. Ensure your Supabase project has sufficient disk capacity to accommodate the imported data. If you know how big your database is going to be, you can manually increase the size here: https://supabase.com/dashboard/project/_/settings/database

Reviewer: ```suggestion
Large datasets consume disk space. Ensure your Supabase project has sufficient disk capacity to accommodate the imported data. If you know how big your database is going to be, you can manually increase the size in your [projects database settings](https://supabase.com/dashboard/project/_/settings/database).
```

Read more about [disk management](https://supabase.com/docs/guides/platform/database-size#disk-management).

### 3. Disabling Triggers

When importing large datasets, it's often beneficial to disable triggers temporarily. Triggers can significantly slow down the import process, especially if they involve complex logic or referential integrity checks. After the import, you can re-enable the triggers.

To disable triggers, use the following SQL commands:

```sql
-- Disable triggers on a specific table
ALTER TABLE table_name DISABLE TRIGGER ALL;

-- To re-enable triggers
ALTER TABLE table_name ENABLE TRIGGER ALL;

```

### 4. Building Indices at the End

Indexing is crucial for query performance, but building indices while importing a large dataset can be time-consuming. Consider building or rebuilding indices after the data import is complete. This approach can significantly speed up the import process and reduce the overall time required.

To build an index after the data import:

```sql
-- Create an index on a table
create index index_name on table_name (column_name);
```

> Read more about [Managing Indexes in PostgreSQL.](https://supabase.com/docs/guides/database/postgres/indexes)

Reviewer: ```suggestion
Read more about [Managing Indexes in PostgreSQL](https://supabase.com/docs/guides/database/postgres/indexes).
```

The "." at the end

> ## **Common Pitfalls to Avoid**

Reviewer: No need for `**`
```suggestion
## Common Pitfalls to Avoid
```

- **Not Planning Ahead**: Importing large datasets should be well-planned. Ensure you have enough disk space, account for statement timeouts, and know which tools to use.
- **Not Monitoring Progress**: Keep an eye on the import progress, especially when using custom scripts or tools. Monitor for any errors or issues.

> - **Not Backing Up Data**: Before any major import, create a backup of your existing data to ensure you can recover if something goes wrong. Read more about [Database Backups.](https://supabase.com/docs/guides/platform/backups#frequency-of-backups)

Reviewer: ```suggestion
- **Not Backing Up Data**: Before any major import, create a backup of your existing data to ensure you can recover if something goes wrong. Read more about [Database Backups](https://supabase.com/docs/guides/platform/backups#frequency-of-backups).
```

import Layout from '~/layouts/DefaultGuideLayout'

export const meta = {
  id: 'database',
  title: 'Importing into Supabase',

>   description: 'Importing Large Datasets into Supabase.',
>   subtitle: 'Importing Large Datasets into Supabase.',

Reviewer 2: ```suggestion
  description: 'Importing large datasets into Supabase.',
  subtitle: 'Importing large datasets into Supabase.',
```

  sidebar_label: 'FUNDAMENTALS',
}

> Importing large datasets requires careful planning and execution to ensure the process is efficient, reliable, and doesn't disrupt your application. In this guide, we cover various best practices and techniques for importing datasets, while also highlighting common pitfalls to avoid.

Reviewer 2: ```suggestion
Importing large datasets requires careful planning and execution to ensure the process is efficient, reliable, and doesn't disrupt your application. In this guide, we cover various best practices and techniques for importing datasets, and highlight common pitfalls to avoid.
```

> ## You have multiple options for importing your data into Supabase

Reviewer 2: ```suggestion
## How to import data into Supabase

You have multiple options for importing your data into Supabase:

1. [CSV import via the Supabase dashboard](#option-1-csv-import-via-supabase-dashboard)
2. [Bulk import using `pgloader`](#option-2-bulk-import-using-pgloader)
3. [Using the Postgres `COPY` command](#option-3-using-postgres-copy-command)
4. [Using the Supabase API](#option-4-using-the-supabase-api)
```

Nice to give these upfront so the reader can choose from here

### Option 1: CSV Import via Supabase Dashboard

Supabase dashboard provides a user-friendly way to import data. However, for very large datasets, this method may not be the most efficient choice, given the size limit is 100MB. It's generally better suited for smaller datasets and quick data imports. Consider using alternative methods like pgloader for large-scale data imports.

> - Navigate to the relevant table in the [Table Editor.](https://supabase.com/dashboard/project/_/editor)

Reviewer 3: ```suggestion
- Navigate to the relevant table in the [Table Editor.](/dashboard/project/_/editor)
```

> - Navigate to the relevant table in the [Table Editor.](/dashboard/project/_/editor)
> - Click on “Insert” then choose "Import Data from CSV" and follow the on-screen instructions to upload your CSV file.

Reviewer 2: ```suggestion
1. Navigate to the relevant table in the [Table Editor.](/dashboard/project/_/editor)
1. Click on “Insert” then choose "Import Data from CSV" and follow the on-screen instructions to upload your CSV file.
```

### Option 2: Bulk Import using pgloader

[pgloader](https://pgloader.io/) is a powerful tool for efficiently importing data into a PostgreSQL database that supports a wide range of source database engines, including MySQL and MS SQL.

You can use it in conjunction with Supabase by following these steps:

> 1. Install pgloader on your local machine or a server. For more info, you can refer to the official pgloader installation page available at: https://pgloader.readthedocs.io/en/latest/install.html

Reviewer 2: ```suggestion
1. Install pgloader on your local machine or a server. For more info, you can refer to the [official pgloader installation page](https://pgloader.readthedocs.io/en/latest/install.html).
```

> ```jsx

Reviewer 2: Indentation on these looks off, code blocks should be part of list item:

![image](https://github.com/supabase/supabase/assets/26616127/90e384d3-b8c2-40d8-9bff-2b72f9bb998d)

$ apt-get install pgloader
```

2. Create a configuration file that specifies the source data and the target Supabase database (e.g., config.load).

Here's an example configuration file:

```sql
LOAD DATABASE
    FROM sourcedb://USER:PASSWORD@HOST/SOURCE_DB
    INTO postgres://postgres:password@db.xxxx.supabase.co:6543/postgres
ALTER SCHEMA 'public' OWNER TO 'postgres';
set wal_buffers = '64MB', max_wal_senders = 0, statement_timeout = 0, work_mem to '2GB';
```

Customize the source and Supabase database URL and options to fit your specific use case:

- `wal_buffers`: This parameter is set to '64MB' to allocate 64 megabytes of memory for write-ahead logging buffers. A larger value can help improve write performance by caching more data in memory before writing it to disk. This can be useful during data import operations to speed up the writing of transaction logs.
- `max_wal_senders`: It is set to 0, to disable replication connections. This is done during the data import process to prevent replication-related conflicts and issues.
- `statement_timeout`: The value is set to 0, which means it's disabled, allowing SQL statements to run without a time limit.

> - `work_mem`: It is set to '2GB', allocating 2 GB of memory for query operations. This enhance the performance of complex queries by allowing larger in-memory datasets.

Reviewer: ```suggestion
- `work_mem`: It is set to '2GB', allocating 2 GB of memory for query operations. This enhances the performance of complex queries by allowing larger in-memory datasets.
```

3. Run pgloader with the configuration file.

```jsx
pgloader config.load
```

For databases using the Postgres engine, we recommend using the [pg_dump](https://www.postgresql.org/docs/current/app-pgdump.html) and [psql](https://www.postgresql.org/docs/current/app-psql.html) command line tools.

### Option 3: Using Postgres Copy command

> Read more about [Bulk data loading.](https://supabase.com/docs/guides/database/tables#bulk-data-loading)

Reviewer 3: Best to use relative links so that you stay within the same environment (eg. preview vs. prod).

```suggestion
Read more about [Bulk data loading.](/docs/guides/database/tables#bulk-data-loading)
```

### Option 4: Using the Supabase API

> The Supabase API allows you to programmatically import data into your tables. You can use various client libraries to interact with the API and perform data import operations. This approach is useful when you need to automate data imports, and it gives you fine-grained control over the process. Refer to our [API guide](https://supabase.com/docs/guides/api) for more details.

Reviewer 3: ```suggestion
The Supabase API allows you to programmatically import data into your tables. You can use various client libraries to interact with the API and perform data import operations. This approach is useful when you need to automate data imports, and it gives you fine-grained control over the process. Refer to our [API guide](/docs/guides/api) for more details.
```

<Admonition type="note">
When importing data via the Supabase API, it's advisable to refrain from bulk imports. This helps ensure a smooth data transfer process and prevents any potential disruptions.

> Read more about [Rate Limiting, Resource Allocation, & Abuse Prevention.](https://supabase.com/docs/guides/platform/going-into-prod#rate-limiting-resource-allocation--abuse-prevention)

Reviewer 3: ```suggestion
Read more about [Rate Limiting, Resource Allocation, & Abuse Prevention.](/docs/guides/platform/going-into-prod#rate-limiting-resource-allocation--abuse-prevention)
```

</Admonition>

> ## Before you dive into the import process, it's essential to plan and prepare

Reviewer 2: ```suggestion
## Preparing to import data

Before importing data, you should have a plan for timing, disk space, and more. Prepare your database in advance. 
```

Better to have short, scannable titles and add any necessary intro as text below

### 1. Statement Timeouts

By default, Supabase enforces query statement timeouts to ensure fair resource allocation and prevent long-running queries from affecting the overall system. When importing large datasets, you may encounter timeouts. To address this:

> - **Increase the Statement Timeout**: You can adjust the statement timeout for your session or connection to accommodate longer-running queries. Be cautious when doing this, as excessively long queries can negatively impact system performance. Read more about [Statement Timeouts](https://supabase.com/docs/guides/database/postgres/configuration).

Reviewer 3: ```suggestion
- **Increase the Statement Timeout**: You can adjust the statement timeout for your session or connection to accommodate longer-running queries. Be cautious when doing this, as excessively long queries can negatively impact system performance. Read more about [Statement Timeouts](/docs/guides/database/postgres/configuration).
```

### 2. Disk Size Considerations

> Large datasets consume disk space. Ensure your Supabase project has sufficient disk capacity to accommodate the imported data. If you know how big your database is going to be, you can manually increase the size in your [projects database settings](https://supabase.com/dashboard/project/_/settings/database).

Reviewer 3: ```suggestion
Large datasets consume disk space. Ensure your Supabase project has sufficient disk capacity to accommodate the imported data. If you know how big your database is going to be, you can manually increase the size in your [projects database settings](/dashboard/project/_/settings/database).
```

> Read more about [disk management](https://supabase.com/docs/guides/platform/database-size#disk-management).

Reviewer 3: ```suggestion
Read more about [disk management](/docs/guides/platform/database-size#disk-management).
```

### 3. Disabling Triggers

When importing large datasets, it's often beneficial to disable triggers temporarily. Triggers can significantly slow down the import process, especially if they involve complex logic or referential integrity checks. After the import, you can re-enable the triggers.

To disable triggers, use the following SQL commands:

```sql
-- Disable triggers on a specific table
ALTER TABLE table_name DISABLE TRIGGER ALL;

-- To re-enable triggers
ALTER TABLE table_name ENABLE TRIGGER ALL;
```

### 4. Building Indices at the End

Indexing is crucial for query performance, but building indices while importing a large dataset can be time-consuming. Consider building or rebuilding indices after the data import is complete. This approach can significantly speed up the import process and reduce the overall time required.

To build an index after the data import:

```sql
-- Create an index on a table
create index index_name on table_name (column_name);
```

> Read more about [Managing Indexes in PostgreSQL](https://supabase.com/docs/guides/database/postgres/indexes).

Reviewer 3: ```suggestion
Read more about [Managing Indexes in PostgreSQL](/docs/guides/database/postgres/indexes).
```

## Common Pitfalls to Avoid

- **Not Planning Ahead**: Importing large datasets should be well-planned. Ensure you have enough disk space, account for statement timeouts, and know which tools to use.
- **Not Monitoring Progress**: Keep an eye on the import progress, especially when using custom scripts or tools. Monitor for any errors or issues.

> - **Not Backing Up Data**: Before any major import, create a backup of your existing data to ensure you can recover if something goes wrong. Read more about [Database Backups](https://supabase.com/docs/guides/platform/backups#frequency-of-backups).

Reviewer 3: ```suggestion
- **Not Backing Up Data**: Before any major import, create a backup of your existing data to ensure you can recover if something goes wrong. Read more about [Database Backups](/docs/guides/platform/backups#frequency-of-backups).
```
